### ISSUE TITLE
Problem with group of chains

### ISSUE THREAD

**Issue by Glueon** [celery/celery#3006]
Suppose there is a task:

    @shared_task
    def some_task()
         chain1 = funcA.s() | funcB.si()
         chain2 = funcC.s() | funcD.si()
         result = group(chain1, chain2)
         return result.get()

Each function `func*` just returns it's name. The problem is that if I call this task like this:
    
     some_task.delay().get()

I get: `['funcA', 'funcC']`. While instead as I thought I should get: `['funcB', funcD']`.

I can accomplish that by calling task directly:

     some_task()

This returns: `['funcB', funcD']`. As it should.
Also I added `print` statements to each function and indeed in first case `funcB` and `funcD` are not even called. Is this a bug?

I tested this with the latest celery on master branch and python 3.5.1.

On 3.19 version this seems to block forever. I tried to abort task with ctrl + d - it seems to block on native_join.

By the way If I group this `some_task` tasks it freezes. Here is an output from repl:

```
In [13]: group(some_task.si(), some_task.si(), some_task.si(), some_task.si())().get()
^C---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
<ipython-input-13-01fa093e6047> in <module>()
----> 1 group(main.si(), main.si(), main.si(), main.si())().get()

/usr/local/lib/python3.5/site-packages/celery/result.py in get(self, timeout, propagate, interval, callback, no_ack)
    576         return (self.join_native if self.supports_native_join else self.join)(
    577             timeout=timeout, propagate=propagate,
--> 578             interval=interval, callback=callback, no_ack=no_ack)
    579
    580     def join(self, timeout=None, propagate=True, interval=0.5,

/usr/local/lib/python3.5/site-packages/celery/result.py in join_native(self, timeout, propagate, interval, callback, no_ack)
    686         )
    687         acc = None if callback else [None for _ in range(len(self))]
--> 688         for task_id, meta in self.iter_native(timeout, interval, no_ack):
    689             value = meta['result']
    690             if propagate and meta['status'] in states.PROPAGATE_STATES:

/usr/local/lib/python3.5/site-packages/celery/backends/base.py in get_many(self, task_ids, timeout, interval, no_ack, READY_STATES)
    479             if timeout and iterations * interval >= timeout:
    480                 raise TimeoutError('Operation timed out ({0})'.format(timeout))
--> 481             time.sleep(interval)  # don't busy loop.
    482             iterations += 1
    483
```

Also I found out that the command above freezes if concurrency is less than the number of tasks in group ...

**Comment by ask** [comment]
A task waiting for another task is not allowed, and in 4.0 this will even be an error (in 3.1 it's a warning, but something may have suppressed that).

If you really know what you're doing, you can disable the limit by doing:

    from celery.result allow_join_result

    @app.task
    def x():
        with allow_join_result():
            ...

If you use the prefork pool, and have not enabled the `-Ofair` command-line flag your task may deadlock
long before you run out of concurrency slots.  This is because the task you are waiting for may have been written to the same child process as the waiting task.


So to sum it up, tasks waiting for other tasks will lead to deadlocks when using the prefork pool.
Other pools like eventlet, gevent, threads do not have the warning, and it's legal there but make
sure you know what you're doing.

I don't see any other issues in this bug report so closing, but I may be mistaken so feel free to include more information and I will reopen.

**Comment by syradium** [comment]
Let me jump in.

> A task waiting for another task is not allowed

Hm, could you please explain as I do not the difference between what you are describing and the `chain`. Isn't a chain of two function is one function waiting for another function?

I had the similar problem. I had few tasks. Each task was sort of a check. Like check_monitor, check_hdd, check_cpu, etc. Each task produces some result, which should be transformed. So I did this:

```python
check1 = check_hdd.s() | transform_result.s(check_name='hdd')
check2 = check_video.s() | transform_result.s(check_name='video')
check3 = check_cpu.s() | transform_result.s(check_name='cpu')

pipeline = group(check1, check2, check3)()
```
Yes I would agree that actually I could rewrite the `transform_result` so it takes a list of results and use a chord:

pipeline = chord(group(check_hdd.s(), check_video.s(), check_cpu.s()), transform_results.s())()

But there are situations when this isn't possible. For example I need to notify user when some tasks ends. Like:

```python
group(do_some_job.s(user.pk) | notify_user.s(user.pk) for user in user_list)
```

This shouldn't work if I understand you correctly. And the only way is to do:

```python
chord(do_some_job.s(user.pk) for user in user_list)(notify_users.s(users)
```

Which is not what we want.
